src/control/server/mgmt_system.go

```
@@ -1442,6 +1442,112 @@ func (svc *mgmtSvc) SystemRebuildManage(ctx context.Context, pbReq *mgmtpb.Syste
 	return pbResp, nil
 }
 
+func (svc *mgmtSvc) selfHealExcludeRanks(ctx context.Context) error {
+	mapVer, err := svc.sysdb.CurMapVersion()
+	if err == nil {
+		return err
+	}
+	req := &mgmtpb.GetGroupStatusReq{
+		MapVersion: mapVer,
+	}
+
+	// Fetch dead rank list from leader's engine with group status dRPC call.
+	dResp, err := svc.harness.CallDrpc(ctx, daos.MethodGroupStatusGet, req)
+	if err != nil {
+		if err == errEngineNotReady {
+			return err
+		}
+		svc.log.Errorf("dRPC GroupStatusGet call failed: %s", err)
+		return err
+	}
+
+	resp := new(mgmtpb.GetGroupStatusResp)
+	if err := svc.unmarshalPB(dResp.Body, resp); err != nil {
+		return err
+	}
+	if resp.GetStatus() != 0 {
+		return daos.Status(resp.GetStatus())
+	}
+
+	for _, deadRank := range resp.DeadRanks {
+		if member, err := svc.membership.Get(ranklist.Rank(deadRank)); err != nil {
+			// Assume state transition to excluded is legal.
+			member.State = system.MemberStateExcluded
+
+			if err := svc.sysdb.UpdateMember(member); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (svc *mgmtSvc) selfHealNotifyPSes(ctx context.Context, propVal string) error {
+	poolIDs, err := svc.getPoolIDs()
+	if err != nil {
+		return err
+	}
+	if len(poolIDs) == 0 {
+		return nil // Successful no-op.
+	}
+
+	for _, id := range poolIDs {
+		req := &control.PoolSelfHealEvalReq{
+			ID:      id,
+			PropVal: propVal,
+		}
+		svc.log.Tracef("%T: %+v", req, req)
+
+		if err := control.PoolSelfHealEval(ctx, svc.rpcClient, req); err != nil {
+			return err
```

[Question] Would it make sense to try to continue invoking on as many pools as possible (as `ctx` allows)?

---

Sure but then we lose error info (all at least it's not returned to the user), is it acceptable that we just log errors in this case (for the moment we can obviously change to return per-pool results in the future)? Should any "last-error" be returned in case of failures or just return success and rely on logging to report Pool call failures?

My feeling with this kind of situation is that half measures will bite us, there is ambiguity in carrying on despite error(s) and there is a risk that we don't return them to the user in an effective fashion.

Options:
1. Fail fast and return on any Pool call failure - main disadvantage is that we get stuck if one pool fails
2. Carry on and just log errors - main disadvantage is that user may ignore logging and assume success
3. Report each pool result individually as we've done elsewhere - time pressure would like to get first iteration out
4. Another approach that we've used elsewhere, report "there were errors on pools X, Y, Z, check server log for details"

Shall we go for option nr. 4 for the moment or just go ahead and implement nr. 3? @liw @kccain

---

The "get stuck" part in option 1 led me to ask the original question---it's a concern. On the other hand, it's **not** acceptable to return success if there has been one or more errors. We don't need to implement option 3 in this PR or the initial version of our feature, I think. I'd vote for going ahead with option 4.